qualtran/qref_interop/_bloq_to_qref.py

```diff
@@ -0,0 +1,311 @@
+        yield (f"out_{reg.name}", "output")
+    if reg.side != Side.RIGHT:
+        yield (f"in_{reg.name}", "input")
+
+
+def _expand_name_if_needed(reg_name, shape) -> Iterable[str]:
+    """Given a register name, expand it into sequence of names if it has nontrivial shape.
+
+    Examples:
+        "reg", () -> "reg"
+        "reg", (3,) -> "reg_0", "reg_1", "reg_2"
+    """
+    return (reg_name,) if shape == () else (f"{reg_name}_{i}" for i in range(shape[0]))
+
+
+def _ports_from_register(reg: Register) -> Iterable[PortV1]:
+    """Given a Qualtran register, return iterable of corresponding Bartiq Ports.
+
+    Intuitively, one would expect a one to one correspondence between ports and registers.
+    However:
+        - We currently don't support bidirectional ports. Hence, THRU registers have to be
+          split into pairs of input/output ports
+        - We currently don't support composite ports. Hence composite registers (i.e. ones
+          with a nontrivial shape) have to be split into multiple ports.
+
+    Args:
+        reg: Register to be converted into ports.
+
+    Raises:
+       NotImplementedError: if `reg` is a compound register with more than one-dimension.
+       Currently we don't support such scenario.
+    """
+    if len(reg.shape) > 1:
+        raise NotImplementedError(
+            "Registers with two or more dimensional shape are not yet supported. "
+            f"The error was caused by register {reg}."
+        )
+
+    return sorted(
+        [
+            # Observe two loops:
+            # - first one splits (if needed) any THRU register into two ports. It also takes care of
+            #   correct naming based on port directions.
+            # - second one expands composite register (which have no counterpart in Bartiq) into
+            #   required number of single ports.
+            PortV1(
+                name=expanded_name,
+                direction=direction,
+                size=reg.bitsize if isinstance(reg.bitsize, int) else str(reg.bitsize),
+            )
+            for flat_name, direction in _names_and_dir_from_register(reg)
+            for expanded_name in _expand_name_if_needed(flat_name, reg.shape)
+        ],
+        key=lambda p: p.name,
+    )
+
+
+def _opposite(direction: str) -> str:
+    return "out" if direction == "in" else "in"
+
+
+def _relative_port_name(soquet: Soquet, direction) -> str:
+    """Given a Soquet and direction, determine the relative name of corresponding Bartiq Port.
+
+    The relative name is always computed wrt. the parent RoutineV1.
+    This function correctly recognizes the fact, that in any connection, the input parent
+    ports serve as outputs for the connection (and vice versa).
+    """
+    if len(soquet.idx) > 1:
+        raise NotImplementedError(
+            "Soquets referencing more than one index in composite register are not yet supported. "
+            f"The error was caused by the following soquet: {soquet}."
+        )
```

Comment:
This answers my question about recursively expanding registers with `len(shape) > 1` -- I think it's worth raising a `NotImplementedError` in `_expand_name_if_needed` method as well.